README.md


# cvar

[![CRANStatusBadge](https://www.r-pkg.org/badges/version/cvar)](https://cran.r-project.org/package=cvar)
[![Build Status](https://travis-ci.com/GeoBosh/cvar.svg?branch=master)](https://travis-ci.com/GeoBosh/cvar)
[![Coverage Status](https://coveralls.io/repos/github/GeoBosh/cvar/badge.svg?branch=master)](https://coveralls.io/github/GeoBosh/cvar?branch=master)


## Overview

Compute expected shortfall (ES) and Value at Risk (VaR) from a
quantile function, distribution function, random number generator or
probability density function.  ES is also known as Conditional Value
at Risk (CVaR). Virtually any continuous distribution can be
specified.  The functions are vectorised over the arguments.
The computations are done directly from the definitions, see e.g. Acerbi
and Tasche (2002). Some support for GARCH models is provided, as well.


## Installing cvar

The [latest stable version](https://cran.r-project.org/package=cvar) is on CRAN. 

    install_packages("cvar")

The vignette shipping with the package gives illustrative examples
(`vignette("Guide_cvar", package = "cvar")`).

You can install the [development version](https://github.com/GeoBosh/cvar) of `cvar` from Github:

    library(devtools)
    install_github("GeoBosh/cvar")


## Overview

Package `cvar` is a small `R` package with, essentially two
functions &#x2014; `ES` for computing the expected shortfall
and `VaR` for Value at Risk.  The user specifies the
distribution by supplying one of the functions that define a
continuous distribution&#x2014;currently this can be a quantile
function (qf), cumulative distribution function (cdf) or
probability density function (pdf). Virtually any continuous
distribution can be specified.

The functions are vectorised over the parameters of the
distributions, making bulk computations more convenient, for
example for forecasting or model evaluation.

The name of this package, "cvar", comes from *Conditional Value at
Risk* (CVaR), which is an alternative term for expected shortfall.

We chose to use the standard names `ES` and `VaR`,
despite the possibility for name clashes with same named
functions in other packages, rather than invent possibly
difficult to remember alternatives. Just call the functions as
`cvar::ES` and `cvar::VaR` if necessary.

Locations-scale transformations can be specified separately
from the other distribution parameters. This is useful when
such parameters are not provided directly by the distribution
at hand. The use of these parameters often leads to more
efficient computations and better numerical accuracy even if
the distribution has its own parameters for this purpose. Some
of the examples for `VaR` and `ES` illustrate this
for the Gaussian distribution.

Since VaR is a quantile, functions computing it for a given
distribution are convenience functions. `VaR` exported by
`cvar` could be attractive in certain workflows because of
its vectorised distribution parameters, the location-scale
transformation and the possibility to compute it from cdf's
when quantile functions are not available.



NEWS.md
# cvar 0.5 (CRAN)

* made `ES` generic (`VaR` was already generic).

* moved `fGarch` from Imports to Suggests.

* renamed argument `x` of `VaR` and `ES` to `p_loss`. `p_loss` seems more
  expressive and suggests that it relates to the losses, usually small numbers
  like `0.05`. Other suitable names like `alpha`, `p`, and `prob`, are commonly
  used as arguments to other functions that might be used as argument `dist` and
  make them mmore difficult to pass via the `...` arguments.

  For now, an warning is issued if `x` is used as a named argument in a call
  (e.g. `VaR(dist, x = 0.05)`) with the intend to turn that in an error in the
  next release of the package. This change should not be noticed by most users
  since it is much more natural not to name this argument and use something like
  `VaR(dist, 0.05)`.

* moved `fGarch` from Imports to Suggests.


# cvar 0.4.1 (CRAN)

* when the input was numeric, `ES()` was not handling the level `x` properly
  (fixes issue #2, reported by Marius Bommert).

* changed the JSS reference to use the new-style doi.

* fixed a bug in the tests, in v0.4-0, that was causing faiure of the tests on
  travis, despite all checks on CRAN passing with OK. `devtools::test()` was
  failing too, but only on the first run in a session, details in the git
  commit.

* set up GHA.


# cvar 0.4-0 (CRAN)

* fix tests to pass with the changed R random generator.

* some new examples and minor documentation changes.


# cvar 0.3-0 (CRAN)

* now `\VignetteIndexEntry` in `Guide_cvar.Rnw` is plain text.

* added experimental support for GARCH models - currently GARCH(1,1) (the API
  may change).

* now the first argument of `VaR()` and `ES()` can be a numeric vector. This is
  useful, e.g., for computing VaR by simulation.

* bugfix:  in `VaR_cdf()` and `VaR_qf()`,  the code for the `if/else` clauses
  had been wrongly swappped. 


# cvar 0.2-0 (CRAN)

* suggest 'covr'.

* setup for travisCI and Coveralls.

* more tests.

* corrected \VignetteIndexEntry in the vignette --- I used the vignette for Rdpack as a
  template but didn't change this entry, which resulted in the vignette appearing with a
  puzzling title on CRAN and other sites.!

* added author@R in DESCRIPTION.

* added experimental web site (docs/)


# cvar 0.1-1 (CRAN)

* added a doi to DESCRIPTION.


# cvar 0.1-0

* first public release.


tests/testthat/test-VaR.R
library(cvar)
context("VaR")

test_that("VaR works ok", {

muA <- 0.006408553
sigma2A <- 0.0004018977

## with quantile function
res1 <- cvar::VaR(qnorm, x = 0.05, mean = muA, sd = sqrt(sigma2A))
res2 <- cvar::VaR(qnorm, x = 0.05, intercept = muA, slope = sqrt(sigma2A))
abs((res2 - res1)) # 0, intercept/slope equivalent to mean/sd

expect_error(cvar::VaR(dnorm, dist.type = "pdf", x = 0.05, intercept = muA, slope = sqrt(sigma2A)), "Not ready")
cvar::VaR_qf(qnorm, x = 0.05, mean = muA, sd = sqrt(sigma2A))

## with cdf the precision depends on solving an equation
res1a <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf", mean = muA, sd = sqrt(sigma2A))
res2a <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf", intercept = muA, slope = sqrt(sigma2A))
abs((res1a - res2)) # 3.287939e-09
abs((res2a - res2)) # 5.331195e-11, intercept/slope better numerically

expect_equal(res1, VaR_qf(qnorm, x = 0.05, mean = muA, sd = sqrt(sigma2A)))
expect_equal(res2a, VaR_cdf(pnorm, x = 0.05, mean = muA, sd = sqrt(sigma2A)))

set.seed(1236)
a.num <- rnorm(100)
VaR(a.num)
ES(a.num)


## as above, but increase the precision, this is probably excessive
res1b <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf", mean = muA, sd = sqrt(sigma2A), tol = .Machine$double.eps^0.75)
res2b <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf", intercept = muA, slope = sqrt(sigma2A), tol = .Machine$double.eps^0.75)

expect_lt( abs((res1b - res2)), 1e-12) # 6.938894e-18 # both within machine precision
expect_lt( abs((res2b - res2)), 1e-12) # 1.040834e-16

## relative precision is also good
abs((res1b - res2)/res2) # 2.6119e-16 # both within machine precision
abs((res2b - res2)/res2) # 3.91785e-15


## an extended example with vector args, if "PerformanceAnalytics" is present
if(require("PerformanceAnalytics")){
    data(edhec, package = "PerformanceAnalytics")
    mu <- apply(edhec, 2, mean)
    sigma2 <- apply(edhec, 2, var)
    musigma2 <- cbind(mu, sigma2)

    ## compute in 2 ways with cvar::VaR
    vAz1 <- cvar::VaR(qnorm, x = 0.05, mean = mu, sd = sqrt(sigma2))
    vAz2 <- cvar::VaR(qnorm, x = 0.05, intercept = mu, slope = sqrt(sigma2))

    vAz1a <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf",
                       mean = mu, sd = sqrt(sigma2))
    vAz2a <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf",
                       intercept = mu, slope = sqrt(sigma2))

    vAz1b <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf",
                   mean = mu, sd = sqrt(sigma2),
                   tol = .Machine$double.eps^0.75)
    vAz2b <- cvar::VaR(pnorm, x = 0.05, dist.type = "cdf",
                   intercept = mu, slope = sqrt(sigma2),
                   tol = .Machine$double.eps^0.75)

    ## analogous calc. with PerformanceAnalytics::VaR
    vPA <- apply(musigma2, 1, function(x)
        PerformanceAnalytics::VaR(p = .95, method = "gaussian", invert = FALSE,
                                  mu = x[1], sigma = x[2], weights = 1))
    ## the results are numerically the same
    max(abs((vPA - vAz1))) # 5.551115e-17
    max(abs((vPA - vAz2))) #   ""

    max(abs((vPA - vAz1a))) # 3.287941e-09
    max(abs((vPA - vAz2a))) #  1.465251e-10, intercept/slope better

    max(abs((vPA - vAz1b))) # 4.374869e-13
    max(abs((vPA - vAz2b))) # 3.330669e-16
}

})

test_that("ES works ok", {

expect_equal(ES(qnorm, dist.type = "qf"),
             ES(qnorm) )

## Gaussian
expect_equal(ES(qnorm, dist.type = "qf"),
             ES(pnorm, dist.type = "cdf") )


## t-dist
expect_equal(ES(qt, dist.type = "qf", df = 4),
             ES(pt, dist.type = "cdf", df = 4) )

expect_equal(ES(pnorm, x = 0.95, dist.type = "cdf"),
             ES(qnorm, x = 0.95, dist.type = "qf") )
## - VaRES::esnormal(0.95, 0, 1)
## - PerformanceAnalytics::ETL(p=0.05, method = "gaussian", mu = 0,
##                             sigma = 1, weights = 1)             # same


## this uses "pdf"
expect_equal(ES(dnorm, x = 0.05, dist.type = "pdf", qf = qnorm),
             ES(pnorm, x = 0.05, dist.type = "cdf") )

})



tests/testthat/test-garch.R
library(cvar)
context("garch")

test_that("GarchModel works ok", {
    mo1a <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5)
    expect_equal(class(mo1a), "garch1c1")

    mo1b <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5, cond.dist = "norm")

    identical(GarchModel(mo1a, omega = 0.4),
              GarchModel(      omega = 0.4, alpha = 0.3, beta = 0.5) )

    ##  GARCH(1,1) with standardised-t_5
    mo_t <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5, cond.dist = list("std", nu = 5))
    expect_identical(mo_t, GarchModel(mo1a, cond.dist = list("std", nu = 5)))
})

test_that("garch1c1 related functions work ok", {
    a_mo <- GarchModel(omega = 0.4, alpha = 0.3, beta = 0.5)
    a <- sim_garch1c1(a_mo, n = 100, n.start = 100, seed = 1234)
    a_pred <- predict(a_mo, n.ahead = 5, Nsim = 100, eps = a$eps, sigmasq = a$h, seed = 1235)

    expect_equal_to_reference(a, "a.RDS")
    expect_equal_to_reference(a_pred, "a_pred.RDS")


    expect_equal(.rgen(NULL), .dist$norm$r)
    ## for now
    expect_error(.rgen("unknowndist"))
    expect_error(.rgen(list("unknowndist")))
    expect_error(.rgen(5))

    expect_equal(.get_cond_dist(NULL, "p"), .dist$norm[["p"]])
    ## for now
    expect_error(.get_cond_dist("unknowndist"))
    expect_error(.get_cond_dist(list("unknowndist")))
    expect_error(.get_cond_dist(5))

})


R/cvar-package.R
#' @author Georgi N. Boshnakov
#'
#' @importFrom stats quantile runif
#' @importFrom gbutils cdf2quantile
#' @importFrom Rdpack reprompt
#'
#' @importFrom fGarch dstd pstd qstd rstd dged pged qged rged
#'
#' @docType package
#' @name cvar-package
#' @aliases cvar
#'
#' @title Compute Conditional Value-at-Risk and Value-at-Risk
#'
#' @description Compute expected shortfall (ES) and Value at Risk (VaR)
#'     from a quantile function, distribution function, random number
#'     generator or probability density function.  ES is also known as
#'     Conditional Value at Risk (CVaR). Virtually any continuous
#'     distribution can be specified.  The functions are vectorised over
#'     the arguments. Some support for GARCH models is provided, as well.
#'
#' @details
#'
#'     There is a huge number of functions for computations with
#'     distributions in core \R and in contributed packages. Pdf's,
#'     cdf's, quantile functions and random number generators are
#'     covered comprehensively. The coverage of expected shortfall is
#'     more patchy but a large collection of distributions, including
#'     functions for expected shortfall, is provided by
#'     \insertCite{VaRES2013;textual}{cvar}.
#'     \insertCite{PerformanceAnalytics2018;textual}{cvar} and
#'     \insertCite{actuarJSS2008;textual}{cvar} provide packages
#'     covering comprehensively various aspects of risk measurement,
#'     including some functions for expected shortfall.
#'
#'     Package \pkg{cvar} is a small package with, essentially, two main
#'     functions --- \code{ES} for computing the expected shortfall
#'     and \code{VaR} for Value at Risk.  The user specifies the
#'     distribution by supplying one of the functions that define a
#'     continuous distribution---currently this can be a quantile
#'     function (qf), cumulative distribution function (cdf) or
#'     probability density function (pdf). Virtually any continuous
#'     distribution can be specified.
#'
#'     The functions are vectorised over the parameters of the
#'     distributions, making bulk computations more convenient, for
#'     example for forecasting or model evaluation.
#'
#'     The name of this package, "cvar", comes from \emph{Conditional
#'     Value at Risk} (CVaR), which is an alternative term for
#'     expected shortfall.
#'
#'     We chose to use the standard names \code{ES} and \code{VaR},
#'     despite the possibility for name clashes with same named
#'     functions in other packages, rather than invent possibly
#'     difficult to remember alternatives. Just call the functions as
#'     \code{cvar::ES} and \code{cvar::VaR} if necessary.
#'
#'     Locations-scale transformations can be specified separately
#'     from the other distribution parameters. This is useful when
#'     such parameters are not provided directly by the distribution
#'     at hand. The use of these parameters often leads to more
#'     efficient computations and better numerical accuracy even if
#'     the distribution has its own parameters for this purpose. Some
#'     of the examples for \code{VaR} and \code{ES} illustrate this
#'     for the Gaussian distribution.
#'
#'     Since VaR is a quantile, functions computing it for a given
#'     distribution are convenience functions. \code{VaR} exported by
#'     \pkg{cvar} could be attractive in certain workflows because of
#'     its vectorised distribution parameters, the location-scale
#'     transformation, and the possibility to compute it from cdf's
#'     when quantile functions are not available.
#'
#'     Some support for GARCH models is provided, as well. It is
#'     currently under development, see \code{\link{predict.garch1c1}}
#'     for current functionality.
#'
#' @references
#'     \insertAllCited{}
#'
#' ---
#'
#' This is for pkgdown test. The following two phrases should be in separate paragraphs:
#'
#' \Sexpr[stage=build,results=rd]{paste0(c("First phrase.", "Second phrase."), collapse = "\n\n")}
#'
#' ---
#'
#' \Sexpr[stage=build,results=rd]{paste0("First phrase.\n\nSecond phrase.")}
#'
#' @examples
#' ## see the examples for ES(), VaR(), predict.garch1c1()
#'
NULL

.onLoad <- function(lib, pkg){
    Rdpack::Rdpack_bibstyles(package = pkg, authors = "LongNames")
    invisible(NULL)
}


vignettes/auto/Guide_cvar.el
(TeX-add-style-hook
 "Guide_cvar"
 (lambda ()
   (TeX-add-to-alist 'LaTeX-provided-class-options
                     '(("jss" "nojss" "article")))
   (TeX-add-to-alist 'LaTeX-provided-package-options
                     '(("fontenc" "T1") ("geometry" "left=2cm" "right=2cm" "bottom=15mm") ("natbib" "authoryear" "round" "longnamesfirst")))
   (add-to-list 'LaTeX-verbatim-environments-local "alltt")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "path")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "url")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "nolinkurl")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "hyperbaseurl")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "hyperimage")
   (add-to-list 'LaTeX-verbatim-macros-with-braces-local "hyperref")
   (add-to-list 'LaTeX-verbatim-macros-with-delims-local "path")
   (TeX-run-style-hooks
    "latex2e"
    "jss"
    "jss10"
    "fontenc"
    "geometry"
    "graphicx"
    "color"
    "alltt"
    "natbib"
    "hyperref"
    "amsmath"
    "amsfonts")
   (TeX-add-symbols
    '("ES" ["argument"] 1)
    '("VaR" ["argument"] 1))
   (LaTeX-add-labels
    "sec:introduction"
    "sec:var"
    "sec:expected-shortfall")
   (LaTeX-add-bibliographies
    "REFERENCES"))
 :latex)



R/garch.R
#' Specify a GARCH model
#'
#' Specify a GARCH model.
#'
#' Argument \code{model} can be the result of a previous call to \code{GarchModel}.
#' Arguments in \code{"..."} overwrite current components of \code{model}.
#'
#' \code{GarchModel} guarantees that code using it will continue to work
#' transparently for the user even if the internal represedtation of GARCH
#' models in this package is changed or additional functionality is added.
#'
#' @param model a GARCH model or a list.
#' @param ... named argument specifying the GARCH model.
#' @param model.class a class for the result. By default \code{GarchModel()}
#'     decides the class of the result.
#' @return an object from suitable GARCH-type class
#'
#' @examples
#' ## GARCH(1,1) with Gaussian innovations
#' mo1a <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5)
#' mo1b <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5, cond.dist = "norm")
#'
#' ## equivalently, the parameters can be given as a list
#' p1 <- list(omega = 1, alpha = 0.3, beta = 0.5)
#' mo1a_alt <- GarchModel(p1)
#' mo1b_alt <- GarchModel(p1, cond.dist = "norm")
#' stopifnot(identical(mo1a, mo1a_alt), identical(mo1b, mo1b_alt))
#'
#' ## additional arguments modify values already in 'model'
#' mo_alt <- GarchModel(p1, beta = 0.4)
#'
#' ## set also initial values
#' mo2 <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5, esp0 = - 1.5, h0 = 4.96)
#'
#' ##  GARCH(1,1) with standardised-t_5
#' mot <- GarchModel(omega = 1, alpha = 0.3, beta = 0.5, cond.dist = list("std", nu = 5))
#'
#' @export
GarchModel <- function(model = list(), ..., model.class = NULL){
    ## TODO: check the correctness of the parameters
    ## alpha, beta, cond.dist, 2nd order stationary, initial vaalue for eps_t^2, h_t

    ## 2019-03-15 TODO: handle fGARCH models
    ## if(inherits(model, "fGARCH")){
    ##
    ## }else{

        dots <- list(...)
        if(length(dots) > 0)
            model[names(dots)] <- dots

        class(model) <- "GarchModel0"
        if(is.null(model.class) && length(model$alpha) == 1 && length(model$beta) == 1)
            class(model) <- "garch1c1"                     # GARCH(1,1), 'c' stands for 'comma'
    ## }

    model
}


.dist <- list(
    norm = list(d = call("dnorm", x = NA, mean = 0, sd = 1),
                p = call("pnorm", q = NA, mean = 0, sd = 1),
                q = call("qnorm", p = NA, mean = 0, sd = 1),
                r = call("rnorm", n = NA, mean = 0, sd = 1)),
    std  = list(d = call("dstd",  x = NA, mean = 0, sd = 1, nu = NA),
                p = call("pstd",  q = NA, mean = 0, sd = 1, nu = NA),
                q = call("qstd",  p = NA, mean = 0, sd = 1, nu = NA),
                r = call("rstd",  n = NA, mean = 0, sd = 1, nu = NA)),
    ged  = list(d = call("dged",  x = NA, mean = 0, sd = 1, nu = NA),
                p = call("pged",  q = NA, mean = 0, sd = 1, nu = NA),
                q = call("qged",  p = NA, mean = 0, sd = 1, nu = NA),
                r = call("rged",  n = NA, mean = 0, sd = 1, nu = NA))
)

.rgen <- function(dist){
    if(is.character(dist)){
        gen <- .dist[[dist]]
        if(is.null(gen)){
            stop("this case is not implemented yet.")
        }else
            return(gen$r)
    }else if(is.null(dist)){
        ## normal distribution is default
        return(.dist$norm$r)
    }else if(is.list(dist)){
        ## assuming dist[[1]] is character string
        gen <- .dist[[ dist[[1]] ]]
        if(is.null(gen)){
            stop("this case is not implemented yet.")
        }else{
            res <- gen$r
            if(length(dist) > 1)  # set parameters if specified
                res[names(dist[-1])] <- dist[-1]

            return(res)
        }
    }

    stop("invalid distribution specification.")
}

.get_cond_dist <- function(dist, what){
   if(is.character(dist)){
        gen <- .dist[[dist]]
        if(is.null(gen)){
            stop("this case is not implemented yet.")
        }else
            return(gen[[what]])
    }else if(is.null(dist)){
        ## normal distribution is default
        return(.dist$norm[[what]])
    }else if(is.list(dist)){
        ## assuming dist[[1]] is character string
        gen <- .dist[[ dist[[1]] ]]
        if(is.null(gen)){
            stop("this case is not implemented yet.")
        }else{
            res <- gen[[what]]
            if(length(dist) > 1)  # set parameters if specified
                res[names(dist[-1])] <- dist[-1]

            return(res)
        }
    }

    stop("this case is not implemented yet.")
}

#' Simulate GARCH(1,1) time series
#'
#' Simulate GARCH(1,1) time series.
#'
#' The simulated time series is in component \code{eps} of the returned value.
#' For exploration of algorithms and eestimation procedures, the volatilities
#' and the standardised innovations are also returned.
#'
#' The random seed at the start of the simulations is saved in the returned
#' object.  A speficific seed can be requested with argument \code{seed}. In
#' that case the simulations are done with the specified seed and the old state
#' of the random number generator is restored before the function returns.
#'
#' @param model a GARCH(1,1) model, an object obtained from \code{GarchModel}.
#' @param n the length of the generated time series.
#' @param n.start number of warm-up values, which are then dropped.
#' @param seed an integer to use for setting the random number generator.
#'
#' @return a list with components:
#' \item{eps}{the time series,}
#' \item{h}{the (squared) volatilities,}
#' \item{eta}{the standardised innovations,}
#' \item{model}{the GARCH(1,1) model,}
#' \item{.sim}{a list containing the parameters of the simulation,}
#' \item{call}{the call.}
#'
#' @note This function is under development and may be changed.
#'
#' @export
sim_garch1c1 <- function(model, n, n.start = 0, seed = NULL){
    ## TODO: seed is not used currently

    garch1c1_fields <- c("omega", "alpha", "beta", "cond.dist", "eps0", "h0", "eps0sq")

    stopifnot(all(names(model) %in% garch1c1_fields))

    omega  <- model$omega
    alpha  <- model$alpha
    beta   <- model$beta
    rgen   <- .rgen(model$cond.dist)

    eps0   <- model$eps0
    h0     <- model$h0
    eps0sq <- model$eps0sq

    ## see Francq & Zakoian, p. 142, for alternative initialisations
    if(is.null(eps0sq)){
        model$eps0sq <- eps0sq <-
            if(is.null(eps0))
                omega / (1 - alpha - beta)
            else
                eps0^2
    }

    if(is.null(h0)){
        model$h0 <- h0 <-
            omega / (1 - alpha - beta)
    }

    N <- n + n.start
    h <- eps <- numeric(N)

    RNGstate <- .RNGstate(seed)
    if(!is.null(RNGstate$oldRNGstate)) # or !is.null(seed)
        on.exit(assign(".Random.seed", RNGstate$oldRNGstate, envir = .GlobalEnv))

    rgen$n <- N # or: rgen[[2]] <- N
    eta <- eval(rgen)

    h[1] <- omega + alpha * eps0sq + beta * h0
    for(i in 1:(N - 1)){
        eps[i] <- sqrt(h[i]) * eta[i]
        h[i + 1] <- omega + alpha * eps[i]^2 + beta * h[i]
    }
    eps[N] <- sqrt(h[N]) * eta[N]

    ## TODO: return eta only conditionally?
    ## TODO: include the model in the returned value?
    ## TODO: return also the initialisation values (in a separate component)?
    list(eps = eps[(n.start + 1):N], h = h[(n.start + 1):N], eta = eta[(n.start + 1):N],
         model = model,
         .sim = list(seed = seed, rand.gen = rgen, n = n, n.start = n.start),
         call = match.call()
         )
}

## modelled after the beginning of `stats:::simulate.lm()`
.RNGstate <- function(seed = NULL){
    if(!exists(".Random.seed", envir = .GlobalEnv, inherits = FALSE))
        runif(1)

    oldRNGstate <- get(".Random.seed", envir = .GlobalEnv)
    if (is.null(seed))
        list(RNGstate = oldRNGstate)
    else{
        ## This doesn't resolve the problem
        ## suppressWarnings(RNGversion("3.5.0"))  # 2019-03-13 temporary, RNG changed in R-devel.
        ##                                        #            see email from Kurt Hornik in Org/
        set.seed(seed)
        RNGstate <- structure(seed, kind = as.list(RNGkind()))
        ## on.exit(assign(".Random.seed", oldRNGstate, envir = .GlobalEnv))
        list(RNGstate = RNGstate, oldRNGstate = oldRNGstate)
    }
}

#' Prediction for GARCH(1,1) time series
#'
#' Predict GARCH(1,1) time series.
#'
#' Plug-in prediction intervals and predictive distributions are obtained by
#' inserting the predicted volatility in the conditional densities. For
#' predictions more than one lag ahead these are not the real predictive
#' distributions but the prediction intervals are usually adequate.
#'
#' For simulation prediction intervals we generate a (large) number of
#' continuations of the given time series. Prediction intervals can be based on
#' sample quantiles. The generated samples are stored in the returned object and
#' can be used for further exploration of the predictive
#' distributions. \code{dist_sim$eps} contains the simulated future values of
#' the time series and \code{dist_sim$h} the corresponding (squared)
#' volatilities.  Both are matrices whose \code{i}-th rows contain the predicted
#' quantities for horizon \code{i}.
#'
#' @param object an object from class \code{"garch1c1"}.
#' @param n.ahead maximum horizon (lead time) for prediction.
#' @param Nsim number of Monte Carlo simulations for simulation based quantities.
#' @param eps the time series to predict, only the last value is used.
#' @param sigmasq the (squared) volatilities, only the last value is used.
#' @param seed an integer, seed for the random number generator.
#' @param ... currently not used.
#'
#' @return an object from S3 class \code{"predict_garch1c1"} containing
#'     the following components:
#' \item{eps}{point predictions (conditional expectations) of the time series (equal
#'     to zero for pure GARCH).}
#' \item{h}{point predictions (conditional expectations)of the squared volatilities.}
#' \item{model}{the model.}
#' \item{call}{the call.}
#' \item{pi_plugin}{Prediction intervals for the time series, based on plug-in
#'     distributions, see Details.}
#' \item{pi_sim}{Simulation based prediction intervals for the time series, see Details.}
#' \item{dist_sim}{simulation samples from the predictive distributions of the time
#'     series and the volatilties. }
#'
#' @examples
#' ## set up a model and simulate a time series
#' mo <- GarchModel(omega = 0.4, alpha = 0.3, beta = 0.5)
#' a1 <- sim_garch1c1(mo, n = 1000, n.start = 100)
#'
#' ## predictions for T+1,...,T+5 (T = time of last value)
#' ## Nsim is small to reduce the load on CRAN, usually Nsim is larger.
#' a.pred <- predict(mo, n.ahead = 5, Nsim = 1000, eps = a1$eps, sigmasq = a1$h, seed = 1234)
#'
#' ## preditions for the time series
#' a.pred$eps
#'
#' ## PI's for eps - plug-in and simulated
#' a.pred$pi_plugin
#' a.pred$pi_sim
#'
#' ## a DIY alculation of PI's using the simulated sample paths
#' t(apply(a.pred$dist_sim$eps, 1, function(x) quantile(x, c(0.025, 0.975))))
#'
#' ## further investigate the predictive distributions
#' t(apply(a.pred$dist_sim$eps, 1, function(x) summary(x)))
#'
#' ## compare predictive densities for h=2 and h=5
#' plot(density(a.pred$dist_sim$eps[2, ]), ylim = c(0,.25))
#' lines(density(a.pred$dist_sim$eps[5, ]), col = "blue")
#'
#' ## predictions of sigma_t^2
#' a.pred$h
#'
#' ## plug-in predictions of sigma_t
#' sqrt(a.pred$h)
#'
#' ## simulation predictive densities of sigma_t for h = 2 and h = 5
#' plot(density(sqrt(a.pred$dist_sim$h[2, ])), xlim = c(0, 6))
#' lines(density(sqrt(a.pred$dist_sim$h[5, ])), col = "blue")
#'
#' ## VaR and ES for different horizons
#' cbind(h = 1:5,
#'       VaR = apply(a.pred$dist_sim$eps, 1, function(x) VaR(x, c(0.05))),
#'       ES = apply(a.pred$dist_sim$eps, 1, function(x) ES(x, c(0.05))) )
#'
#' ## fit a GARCH(1,1) model to exchange rate data and predict
#' gmo1 <- fGarch::garchFit(formula = ~garch(1, 1), data = fGarch::dem2gbp,
#'   include.mean = FALSE, cond.dist = "norm", trace = FALSE)
#' mocoef <- gmo1@fit$par
#' mofitted <- GarchModel(omega = mocoef["omega"], alpha = mocoef["alpha1"],
#'   beta = mocoef["beta1"])
#' gmo1.pred <- predict(mofitted, n.ahead = 5, Nsim = 1000, eps = gmo1@data,
#'   sigmasq = gmo1@h.t, seed = 1234)
#' gmo1.pred$pi_plugin
#' gmo1.pred$pi_sim
#'
#' @note This function is under development and may be changed.
#'
#' @export
predict.garch1c1 <- function(object, n.ahead = 1, Nsim = 1000, eps, sigmasq, seed = NULL, ...){
    model <- object
    omega <- model$omega
    alpha <- model$alpha
    beta  <- model$beta

    model$eps0 <- eps0 <- eps[length(eps)]
    model$h0   <- h0   <- sigmasq[length(sigmasq)] # TODO: compute if not present?

    ## TODO: add non-zero mean and possibly ARMA conditional mean
    pred_eps <- numeric(n.ahead)

    ## tsff slides, p. II.7.4/y1718
    pred_h <- numeric(n.ahead)
    pred_h[1] <- omega + alpha * eps0^2 + beta * h0
    if(n.ahead > 1)
        for(i in 2:n.ahead){
            pred_h[i] <- omega + (alpha + beta) * pred_h[i - 1]
        }

    fq <- .get_cond_dist(model$cond.dist, "q")
    fq$sd <- sqrt(pred_h)

    fq[[2]] <- 0.025
    lwr <- eval(fq)
    fq[[2]] <- 0.975
    upr <- eval(fq)

    eps_pi <- cbind(lwr, upr)

    ## need local variants of withr::with_seed() withr::with_preserve_seed()
    ## but such are not available (as of 2018-09-28), so do it DIY:
    ##
    ## Memorise the state in any case but restore the old state only if 'seed' is not NULL.
    RNGstate <- .RNGstate(seed)
    if(!is.null(RNGstate$oldRNGstate)) # or !is.null(seed)
        on.exit(assign(".Random.seed", RNGstate$oldRNGstate, envir = .GlobalEnv))


    h_sim <- eps_sim <- matrix(NA_real_, nrow = n.ahead, ncol = Nsim)
    for(i in 1:Nsim){
        wrk <- sim_garch1c1(model = model, n = n.ahead, n.start = 0)
        eps_sim[ , i] <- wrk$eps
        h_sim[ , i] <- wrk$h
    }


    sim_pi <- t(apply(eps_sim, 1, function(x) quantile(x, c(0.025, 0.975))))

    structure(list(
        eps = pred_eps, h = pred_h, model = model, call = match.call(),
        pi_plugin = eps_pi, pi_sim = sim_pi,
        dist_sim = list(eps = eps_sim, h = h_sim, RNGstate = RNGstate$RNGstate)
    ), class = "predict_garch1c1")
}





